service/src/main/kotlin/fi/espoo/evaka/reservations/ReservationQueries.kt

```diff
@@ -542,3 +546,224 @@ fun Database.Read.getReservationContractDayRanges(
+    unitId: DaycareId
+): List<DailyChildReservationInfoRow> {
+    return createQuery(
+            """
+with placement_days as
+         (SELECT pl.child_id,
+                 pl.start_date        as placement_start,
+                 pl.end_date          as placement_end,
+                 gpl.start_date       as grouping_start,
+                 gpl.end_date         as grouping_end,
+                 gpl.daycare_group_id as group_id
+          FROM placement pl
+                   JOIN daycare_group_placement gpl
+                             ON pl.id = gpl.daycare_placement_id
+
+          WHERE unit_id = :unitId
+            AND daterange(pl.start_date, pl.end_date, '[]') @> :examinationDate
+            AND daterange(gpl.start_date, gpl.end_date, '[]') @> :examinationDate
+            AND NOT EXISTS (select
+                            from backup_care bc
+                            where bc.child_id = pl.child_id
+                              AND bc.unit_id = pl.unit_id
+                              AND daterange(bc.start_date, bc.end_date, '[]') @> :examinationDate)
+          UNION
+          SELECT bc.child_id,
+                 bc.start_date as placement_start,
+                 bc.end_date   as placement_end,
+                 bc.start_date as grouping_start,
+                 bc.end_date   as grouping_end,
+                 bc.group_id   as group_id
+          FROM backup_care bc
+          WHERE unit_id = :unitId
+            AND daterange(bc.start_date, bc.end_date, '[]') @> :examinationDate)
+SELECT pcd.child_id,
+       p.date_of_birth,
+       p.first_name,
+       p.last_name,
+       p.preferred_name,
+       pcd.group_id,
+       ar.start_time   as reservation_start_time,
+       ar.end_time     as reservation_end_time,
+       ab.absence_type as absence_type,
+       dst.type        as daily_service_time_type,
+       dst.validity_period,
+       dst.regular_times,
+       dst.monday_times,
+       dst.tuesday_times,
+       dst.wednesday_times,
+       dst.thursday_times,
+       dst.friday_times,
+       dst.saturday_times,
+       dst.sunday_times,
+       bc.unit_id      as backup_unit_id
+FROM placement_days pcd
+         JOIN person p ON pcd.child_id = p.id
+         JOIN daycare u ON u.id = :unitId
+         LEFT JOIN absence ab
+                   ON ab.child_id = pcd.child_id
+                       AND ab.date = :examinationDate
+         LEFT JOIN backup_care bc
+                   ON bc.child_id = pcd.child_id
+                       AND bc.unit_id <> :unitId
+                       AND
+                      daterange(bc.start_date, bc.end_date, '[]') @>
+                      :examinationDate
+         LEFT JOIN attendance_reservation ar
+                   ON ar.child_id = pcd.child_id
+                       AND ar.date = :examinationDate
+         LEFT JOIN daily_service_time dst
+                   ON pcd.child_id = dst.child_id
+                       AND dst.validity_period @> :examinationDate
```

Comment:
I think this query would be easier to understand if it used the `realized_placement_one` function which already contains the logic that's in `placement_days` CTE in this query.

I'm also pretty sure that all the reservation and absence edge cases cannot be handled in a single SQL query and will need to be done in the Kotlin side instead. The edge cases have been covered in both citizen calendar and unit's week calendar.

Reply:
Query refactored, but I wasn't really able to figure out a Kotlin side solution that was "better" than the group by summation. Admittedly the deduction logic is perilously complicated but I tried to address this in integration testing.